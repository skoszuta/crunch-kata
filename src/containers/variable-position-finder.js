import _ from 'lodash';
import React from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';

import getVariableName from 'utils/get-variable-name';
import getVariablePath from 'utils/get-variable-path';
import Chip, { ChipsList } from 'components/chip';

const BreadcrumbsWrapper = styled.div`
  margin-top: 1rem;
`;

class VariablePositionFinder extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      selectedVariable: void 0,
      breadcrumbs: []
    };
  }

  onSelectChange({ target: { value }}) {
    const { order } = this.props.variables;

    this.setState({
      selectedVariable: value,
      breadcrumbs: getVariablePath(value, this.props.variables.order)
    });
  }

  renderBreadcrumbs() {
    if (this.state.selectedVariable && !this.state.breadcrumbs) {
      return (
        <div>Variable was not found in the tree.</div>
      );
    } else if (!this.state.selectedVariable) {
      return '';
    }

    const { index } = this.props.variables;

    const breadcrumbs = [
      'Root',
      ...this.state.breadcrumbs,
      getVariableName(index[this.state.selectedVariable])
    ].map((crumb, index, array) => <Chip arrow={index < array.length - 1} key={index}>{crumb}</Chip>);

    return (
      <ChipsList>{breadcrumbs}</ChipsList>
    );
  }

  render() {
    const options = _.map(this.props.variables.index, (variable, key) => (
      <option value={key} key={key}>{getVariableName(variable, key)}</option>
    ));

    if (this.state.selectedVariable === void 0) {
      options.unshift(<option value={void 0} key="placeholder">Select a variable to see its position...</option>);
    }

    return (
      <div>
        <select
          value={this.state.selectedVariable}
          onChange={this.onSelectChange.bind(this)}>
          {options}
        </select>
        <BreadcrumbsWrapper>{this.renderBreadcrumbs()}</BreadcrumbsWrapper>
      </div>
    );
  }
}

function mapStateToProps({ variables }) {
  return { variables };
}

export default connect(mapStateToProps, null)(VariablePositionFinder);
