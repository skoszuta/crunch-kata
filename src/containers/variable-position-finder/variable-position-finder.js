import _ from 'lodash';
import React from 'react';
import { connect } from 'react-redux';

import getVariableName from '../../utils/get-variable-name';
import getVariablePath from '../../utils/get-variable-path';

class VariablePositionFinder extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      selectedVariable: void 0,
      breadcrumbs: []
    };
  }

  onSelectChange({ target: { value }}) {
    const { order } = this.props.variables;

    this.setState({
      selectedVariable: value,
      breadcrumbs: getVariablePath(value, this.props.variables.order)
    });
  }

  renderBreadcrumbs() {
    if (this.state.selectedVariable && !this.state.breadcrumbs) {
      return (
        <div>Variable was not found in the tree.</div>
      );
    } else if (!this.state.selectedVariable) {
      return '';
    }

    const { index } = this.props.variables;

    return this.state.breadcrumbs
      .concat(getVariableName(index[this.state.selectedVariable]))
      .join(' -> ');
  }

  render() {
    const options = _.map(this.props.variables.index, (variable, key) => (
      <option value={key} key={key}>{getVariableName(variable, key)}</option>
    ));

    if (this.state.selectedVariable === void 0) {
      options.unshift(<option value={void 0} key="placeholder">Select a variable to see its position...</option>);
    }

    return (
      <div>
        <select
          value={this.state.selectedVariable}
          onChange={this.onSelectChange.bind(this)}>
          {options}
        </select>
        <div>{this.renderBreadcrumbs()}</div>
      </div>
    );
  }
}

function mapStateToProps({ variables }) {
  return { variables };
}

export default connect(mapStateToProps, null)(VariablePositionFinder);
