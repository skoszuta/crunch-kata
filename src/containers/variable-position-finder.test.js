import fs from 'fs';
import React from 'react';
import { mount, render } from 'enzyme';
import 'jest-styled-components';

import { VariablePositionFinder } from './variable-position-finder';

const { index } = JSON.parse(fs.readFileSync('json/variables.json', 'utf8'));
const { graph: order } = JSON.parse(fs.readFileSync('json/order.json', 'utf8'));

describe('VariablePositionFinder', () => {
  const variables = { index, order };

  describe('snapshot', () => {
    it('should render correctly', () => {
      const wrapper = render(
        <VariablePositionFinder variables={variables} />
      );

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('behavior', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = mount(
        <VariablePositionFinder variables={variables} />
      );
    });

    it('should show a (correct) path', () => {
      expect(wrapper.find('li').length).toBe(0);

      wrapper.find('select').prop('onChange')({ target: { value: 'ede6a8' } });
      wrapper.update();

      expect(wrapper.find('li').length).toBe(7);
    });
  });
});
