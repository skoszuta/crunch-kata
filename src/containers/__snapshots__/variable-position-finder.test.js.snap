// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VariablePositionFinder snapshot should render correctly 1`] = `
.c0 {
  margin-top: 1rem;
}

<div>
  <select>
    <option>
      Select a variable to see its position...
    </option>
    <option
      value="786c0f"
    >
      786c0f - Unaided Awareness (Coffee Roasters)
    </option>
    <option
      value="d88613"
    >
      d88613 - City
    </option>
    <option
      value="0894c5"
    >
      0894c5 - Weight
    </option>
    <option
      value="0323cf"
    >
      0323cf - Wave
    </option>
    <option
      value="d0fe8b"
    >
      d0fe8b - weight
    </option>
    <option
      value="2d27ab"
    >
      2d27ab - Aided Awareness (Coffee Roasters)
    </option>
    <option
      value="7a89e0"
    >
      7a89e0 - Aided Awareness (Taxis)
    </option>
    <option
      value="f2a681"
    >
      f2a681 - Specific Service Usage
    </option>
    <option
      value="8f7db6"
    >
      8f7db6 - Ritual Roasters Usage
    </option>
    <option
      value="e08fbc"
    >
      e08fbc - Education
    </option>
    <option
      value="5b6a73"
    >
      5b6a73 - Children Under 18
    </option>
    <option
      value="ede6a8"
    >
      ede6a8 - Consideration
    </option>
    <option
      value="9f4b60"
    >
      9f4b60 - (unnamed variable)
    </option>
    <option
      value="91a33f"
    >
      91a33f - Family Income
    </option>
    <option
      value="3a5f89"
    >
      3a5f89 - FOURBARREL Usage
    </option>
    <option
      value="62c00f"
    >
      62c00f - Likely Purchase
    </option>
    <option
      value="ee3e40"
    >
      ee3e40 - Stumptown Coffee Roasters Usage
    </option>
    <option
      value="11b0b9"
    >
      11b0b9 - Unaided Awareness (Taxis)
    </option>
    <option
      value="c83da2"
    >
      c83da2 - Gender
    </option>
    <option
      value="1b44a4"
    >
      1b44a4 - Marital Status
    </option>
    <option
      value="0f6ce0"
    >
      0f6ce0 - Usage
    </option>
    <option
      value="8d7127"
    >
      8d7127 - Blue Bottle Coffee Usage
    </option>
  </select>
  <div
    class="c0"
  />
</div>
`;
